app/src/behaviors/behavior_macro.c

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright (c) 2020 The ZMK Contributors
+ *
+ * SPDX-License-Identifier: MIT
+ */
+
+#define DT_DRV_COMPAT zmk_behavior_macro
+
+#include <device.h>
+#include <drivers/behavior.h>
+#include <dt-bindings/zmk/macros.h>
+#include <logging/log.h>
+#include <zmk/behavior.h>
+#include <zmk/behavior_queue.h>
+#include <zmk/keymap.h>
+
+LOG_MODULE_DECLARE(zmk, CONFIG_ZMK_LOG_LEVEL);
+
+#if DT_HAS_COMPAT_STATUS_OKAY(DT_DRV_COMPAT)
+
+enum behavior_macro_mode {
+    MACRO_MODE_TAP,
+    MACRO_MODE_PRESS,
+    MACRO_MODE_RELEASE,
+};
+
+struct behavior_macro_item {
+    uint32_t action;
+    struct zmk_behavior_binding binding;
+};
+
+struct behavior_macro_state {
+    int8_t wait_release_index;
+};
+
+struct behavior_macro_config {
+    uint32_t default_wait_ms;
+    uint32_t default_tap_ms;
+    uint8_t count;
+    struct behavior_macro_item items[];
+};
+
+#define TAP_MODE DT_LABEL(DT_INST(0, zmk_behavior_macro_tap))
+#define PRESS_MODE DT_LABEL(DT_INST(0, zmk_behavior_macro_press))
+#define REL_MODE DT_LABEL(DT_INST(0, zmk_behavior_macro_release))
+
+#define TAP_TIME DT_LABEL(DT_INST(0, zmk_behavior_macro_tap_time))
+#define WAIT_TIME DT_LABEL(DT_INST(0, zmk_behavior_macro_wait_time))
+
+#define WAIT_REL DT_LABEL(DT_INST(0, zmk_behavior_macro_wait_release))
+
+static int behavior_macro_init(const struct device *dev) { return 0; };
+
+static void queue_macro(uint32_t position, const struct behavior_macro_config *cfg,
+                        struct behavior_macro_state *state, bool press) {
+
+    uint32_t wait_ms = cfg->default_wait_ms;
+    uint32_t tap_ms = cfg->default_tap_ms;
+    enum behavior_macro_mode mode = MACRO_MODE_TAP;
+
+    for (int i = 0; i < cfg->count; i++) {
+        if (strcmp(cfg->items[i].binding.behavior_dev, TAP_MODE) == 0) {
+            mode = MACRO_MODE_TAP;
+            LOG_DBG("macro mode set: tap");
+        } else if (strcmp(cfg->items[i].binding.behavior_dev, PRESS_MODE) == 0) {
+            LOG_DBG("macro mode set: press");
+            mode = MACRO_MODE_PRESS;
+        } else if (strcmp(cfg->items[i].binding.behavior_dev, REL_MODE) == 0) {
+            LOG_DBG("macro mode set: release");
+            mode = MACRO_MODE_RELEASE;
+        } else if (press && strcmp(cfg->items[i].binding.behavior_dev, WAIT_REL) == 0) {
+            LOG_DBG("macro: stopping until macro is released");
+            state->wait_release_index = i;
```

Comment:
It might be even simpler to set "wait_release_seen" to `True`:

```
         } else if (strcmp(cfg->items[i].binding.behavior_dev, WAIT_REL) == 0) {
             wait_release_seen = true;
             if(press) { break; }
        }
```
Then change this:
```
         } else if (press || (state->wait_release_index > 0 && i > state->wait_release_index)) {
```
to 
```
         } else if (press || wait_release_seen)) {
```
That way no state needs to be kept at all.

Reply:
I was thinking of a further refinement to actually store the wait/tap milliseconds to the state as well, so that on release, we actually would start our for loop at the previous stop point, avoiding extra unnecessary looping/state changes we've already done on press.

Thoughts?

Reply:
Can the same macro be used on multiple keys? If so, what happens if you press the macro, then while still holding it press it again on a different key? Do the two instances have separate state, or does saving any state to the macro cause it to be overwritten in that case?

Reply:
That would definitely be a limitation of the implementation, I'm not sure if there are any valid use cases for that, that we should worry about it for our first pass of this. Anybody got one?